src/ch/ch_monitor.c

```diff
@@ -997,12 +1022,23 @@ virCHMonitorPostInitialize(virCHMonitorPtr mon, virDomainObjPtr vm,
     if (virDomainObjSave(vm, driver->xmlopt, cfg->stateDir) < 0)
         return -1;
 
+    /*
+     * Opening FIFO will block until the other end is also opened.
+     * This can cause a potential deadlock if done from the main
+     * thread. So we defer the opening to monitor thread. The same
+     * block can happen on cloud-hypervisor as well. So we start the
+     * monitor thread early here and then wait in the thread until
+     * cloud-hypervisor also opens the FIFO.
+     */
+    if (virCHMonitorStartEventLoop(mon) < 0)
+        return -1;
+
     /* get a curl handle */
     mon->handle = curl_easy_init();
 
     /* Try pinging the VMM to make sure it is ready */
```

Comment:
Do we still need to use the ping since we will have the `monitor_fd` and if it opens then we can assume cloud-hypervisor is ready?

Reply:
I feel this is a good to have. Successful opening of monitor file just guarentees that the cloud-hypervisor as a process started and also opened the file. But successful ping is a guarantee that it is ready for accepting API requests.

Reply:
@rbradford could you tell me if the `monitor_fd` is opened by cloud-hypervisor then would the API socket would be able to handle requests? I'd really prefer to not have this ping timeout around if we don't need it (libvirt upstream really disliked it too).

Reply:
What concerns me is the fact that we are relying on a specific behavior of FIFO to judge the health of VMM. Only FIFO has this blocking behavior until the other end is also open. If we decide to use some other mechanism in future(like Unix domain socket etc), we will not have this blocking behavior.

Is it the retry logic that is the problem with upstream or the Ping itself?

Reply:
Upstream is really looking to avoid the retry loop. If the ping blocked until the API responded it wouldn't be a problem I don't think.

Reply:
Oh indeed they are. Hrm let's just keep things as they are for now since this all works and I can poke upstream more when submitting patches again.